tests/Routing/RoutingUrlGeneratorTest.php

```diff
@@ -115,6 +115,64 @@ public function testBasicRouteGeneration()
 	}
 
 
+	public function testBasicRouteGenerationOptionalParameterMatch()
+	{
+		$url = new UrlGenerator(
+				$routes = new Illuminate\Routing\RouteCollection,
+				$request = Illuminate\Http\Request::create('http://www.foo.com/')
+		);
+
+		$route = new Illuminate\Routing\Route(array('GET'), 'foo/bar/{baz?}', array('as' => 'foobaz'));
+		$routes->add($route);
+
+		$this->assertEquals('http://www.foo.com/foo/bar/taylor', $url->route('foobaz', ['baz' => 'taylor']));
+	}
+
+
+	public function testBasicRouteGenerationOptionalParameterNoMatch()
+	{
+		$url = new UrlGenerator(
+			$routes = new Illuminate\Routing\RouteCollection,
+			$request = Illuminate\Http\Request::create('http://www.foo.com/')
+		);
+
+		$route = new Illuminate\Routing\Route(array('GET'), 'foo/bar/{baz?}', array('as' => 'foobaz'));
+		$routes->add($route);
+
+		$this->assertEquals('http://www.foo.com/foo/bar?taylor=otwell', $url->route('foobaz', ['taylor' => 'otwell']));
+	}
+
+
+	public function testBasicRouteGenerationOptionalParameterNoMatchIntegerKeyValue()
+	{
+		$url = new UrlGenerator(
+				$routes = new Illuminate\Routing\RouteCollection,
+				$request = Illuminate\Http\Request::create('http://www.foo.com/')
+		);
+
+		$route = new Illuminate\Routing\Route(array('GET'), 'foo/bar/{baz?}', array('as' => 'foobaz'));
+		$routes->add($route);
+
+		// or /foo/bar?1&2 ???
+		$this->assertEquals('http://www.foo.com/foo/bar//1/2', $url->route('foobaz', [1,2]));
```

Comment:
Well, we'll need a solution that doesn't break backwards compatibility. The
first solution broke when trying to make any URL to a Route::controller
generated route unless you manually specify the optional parameter key.

On Fri, Mar 27, 2015 at 4:20 PM, Aimeos <notifications@github.com> wrote:

> In tests/Routing/RoutingUrlGeneratorTest.php
> <https://github.com/laravel/framework/pull/8198#discussion_r27332043>:
>
> > +		$this->assertEquals('http://www.foo.com/foo/bar?taylor=otwell', $url->route('foobaz', ['taylor' => 'otwell']));
> > +	}
> > +
> > +
> > +	public function testBasicRouteGenerationOptionalParameterNoMatchIntegerKeyValue()
> > +	{
> > +		$url = new UrlGenerator(
> > +				$routes = new Illuminate\Routing\RouteCollection,
> > +				$request = Illuminate\Http\Request::create('http://www.foo.com/')
> > +		);
> > +
> > +		$route = new Illuminate\Routing\Route(array('GET'), 'foo/bar/{baz?}', array('as' => 'foobaz'));
> > +		$routes->add($route);
> > +
> > +		// or /foo/bar?1&2 ???
> > +		$this->assertEquals('http://www.foo.com/foo/bar//1/2', $url->route('foobaz', [1,2]));
>
> @GrahamCampbell <https://github.com/GrahamCampbell> @taylorotwell
> <https://github.com/taylorotwell> I'm open to alternative solutions but
> the current algorithm is broken for optional parameters.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laravel/framework/pull/8198/files#r27332043>.
>